lib/teleterm/daemon/mfaPrompt.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package daemon
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	wancli "github.com/gravitational/teleport/lib/auth/webauthncli"
+	"github.com/gravitational/teleport/lib/client/mfa"
+)
+
+type mfaPrompt struct {
+	clusterURI       string
+	tshdEventsClient api.TshdEventsServiceClient
+	logger           *logrus.Logger
+}
+
+func (s *Service) NewMFAPrompt(clusterURI string) mfa.PromptMFA {
+	return &mfaPrompt{
+		clusterURI:       clusterURI,
+		tshdEventsClient: s.tshdEventsClient,
+		logger:           s.cfg.Log.Logger,
+	}
+}
+
+func (p *mfaPrompt) promptMFA(ctx context.Context, chal *proto.MFAAuthenticateChallenge, cfg mfa.PromptConfig, totp, webauthn bool) (*proto.MFAAuthenticateResponse, error) {
+	resp, err := p.tshdEventsClient.PromptMFA(ctx, &api.PromptMFARequest{
+		RootClusterUri: p.clusterURI,
+		Reason:         cfg.PromptReason,
+		Totp:           totp,
+		Webauthn:       webauthn,
+	})
+	if err != nil {
+		return nil, trail.FromGRPC(err)
+	}
+
+	return &proto.MFAAuthenticateResponse{
+		Response: &proto.MFAAuthenticateResponse_TOTP{
+			TOTP: &proto.TOTPResponse{Code: resp.TotpCode},
+		},
+	}, nil
```

Comment:
Currently, it's not necessary. The flow I laid out here is:

```mermaid
sequenceDiagram
    participant tshd as tsh daemon
    participant electron as Electron app
    actor user
    tshd->>electron: PrompMFA req
    electron->>user: show MFA modal
    
    alt user enters otp
        user->>electron: otp code
        electron->>tshd: otp code
    else user closes modal
        user->>electron: close modal
        electron->>tshd: cancel request
    else user completes webauthn
        user->>tshd: user taps webauthn key
        tshd->>electron: close modal (no error)
    end
```

Previously I had trouble handling tshd cancelling its request without outputting an error in the Electron App, but I'm going to try again. Otherwise we would need to add another call from Electron -> tshd to handle Webauthn, which is similar to how we handled headless. The extra steps in the headless flow made more sense whereas I think we should try to get away with the short flow here.

Reply:
Cancel request and close modal is where my subsequent questions would be. How do you plan to implement "close modal"?

## Close modal

It seems to me that it might be worthwhile to tie this logic to the request-response cycle. In that sense, closing the modal by tsh could be implemented by cancelling the request. I'm not sure what you mean by cancelling a request without outputting an error in the Electron app, but AFAIR the relogin logic already handles request cancellation.

https://github.com/gravitational/teleport/blob/44f17e72886b0c79bc07bdc950a4fc2585a93780/web/packages/teleterm/src/ui/appContext.ts#L172-L178

https://github.com/gravitational/teleport/blob/44f17e72886b0c79bc07bdc950a4fc2585a93780/web/packages/teleterm/src/ui/services/relogin/reloginService.ts#L63

This is hidden behind some layers of abstraction but essentially when the Go code cancels the request, grpc-js emits the `cancelled` event on the `call` object.

https://github.com/gravitational/teleport/blob/44f17e72886b0c79bc07bdc950a4fc2585a93780/web/packages/teleterm/src/services/tshdEvents/index.ts#L151-L161

## Cancel request

Here I suppose the tshd events service could return a response with code ABORTED. I haven't done that anywhere yet, but taking the relogin service just as an example, it'd have to look something like this:

```diff
diff --git a/web/packages/teleterm/src/services/tshdEvents/index.ts b/web/packages/teleterm/src/services/tshdEvents/index.ts
index 8c98badf3e..8a1831e890 100644
--- a/web/packages/teleterm/src/services/tshdEvents/index.ts
+++ b/web/packages/teleterm/src/services/tshdEvents/index.ts
@@ -163,6 +163,18 @@ function createService(logger: Logger): {
           callback(null, new api.ReloginResponse());
         },
         error => {
+          // error passes through the context bridge [1] so none of its properties are preserved
+          // other than the message.
+          //
+          // [1] https://www.electronjs.org/docs/latest/api/context-bridge
+          if (error?.message?.includes('aborted')) {
+            const status = new grpc.StatusBuilder()
+              .withCode(grpc.status.ABORTED)
+              .build();
+            callback(status);
+            return;
+          }
+
           callback(error);
         }
       );
diff --git a/web/packages/teleterm/src/ui/services/relogin/reloginService.ts b/web/packages/teleterm/src/ui/services/relogin/reloginService.ts
index 06aa001018..6036f1bcd6 100644
--- a/web/packages/teleterm/src/ui/services/relogin/reloginService.ts
+++ b/web/packages/teleterm/src/ui/services/relogin/reloginService.ts
@@ -60,6 +60,8 @@ export class ReloginService {
           reject(new Error('Login process was canceled by the user')),
       });
 
+      throw new Error('aborted');
+
       onRequestCancelled(closeDialog);
     });
   }
```